### DISCUSSION TITLE
ggplot text size output differs inline and in knitted HTML

### DISCUSSION THREAD

**Discussion by nikita-telkar** [quarto-dev/quarto-cli#4848] [Q&A]
Hi,

I've been loving using Quarto, but one major issue for me is that the test size in the ggplot figures I see produced inline within my `.qmd` differs from what I see when I knit to HTML. 

This was never an issue in .Rmd. The output I saw inline for `.Rmd`, is what I saw in my HTML i.e., **the text would automatically scale up**  -- how can I retain the same in Quarto documents?

Within my custom theme, I have the following, but I can't keep changing the font sizes per plot (as pretty much all of my plots have varying heights and widths) 

```
my_theme2 <- theme_minimal() +
  theme(plot.title = element_text(family = "Jost", size = 14),
        plot.subtitle = element_text(family = "Jost", size = 12),
        legend.title = element_text(family = "Jost", size = 12),
        legend.text = element_text(family = "Jost", size = 10),
        axis.title = element_text(family = "Jost", size = 14),
        axis.text = element_text(family = "Jost", size = 12),
        strip.text = element_text(family = "Jost", size = 12),
        plot.caption = element_text(family = "Jost", size = 10),
        panel.background = ggplot2::element_rect(color = "#F2F2F2"), 
        strip.background = ggplot2::element_rect(color = "#F2F2F2", fill = "#F2F2F2"),
        panel.spacing = unit(2, "lines"))
```
**e.g., if I set the fig.width = 10 within my code chunk header, I get the two differing ggplots:**

**INLINE:**

![image_inline](https://user-images.githubusercontent.com/59856969/225427847-586e2c23-6b1c-4d00-8d9d-30d8e101c4c5.png)

**HTML:**

![image_html](https://user-images.githubusercontent.com/59856969/225427858-49676b14-9535-4b2a-9aea-ba2003fa6b1d.png)

Thank you very much!

**Comment by baptiste** [comment]
can you please provide a reproducible example, including the settings you use to set the figure size (fig-width and/or out-width)? 
My understanding is that knitr produces plots at a fixed size set by `fig.width` (here 10 inches, so wider than a standard pdf text width, and `out.width` (optionally) to resize it to fit in a given space (which can therefore shrink/expand fonts from the original). And then there is an additional consideration, which is the size of the container in the parent document: in LaTeX, the text width can be varied, and if the plot is made to fill the space it will shrink/expand accordingly. Similarly, in HTML the width is set by the layout you use, and the [figure environment](https://quarto.org/docs/authoring/article-layout.html).

So, in short, it's hard to tell without knowing more precisely what settings you're using.

**Reply by nikita-telkar** [reply]
Thanks so much @baptiste!

Here are all of the options I'm using. I don't have any specified `out.width`, just the `fig.height` and `fig.width` that I keep changing within each chunk header, and just the default `column.body` 

Chunk: {r hsa-miR-196b-5p, fig.width=10, fig.height=6}
Global: `knitr::opts_chunk$set(warning = FALSE, message = FALSE, error = FALSE, cache = TRUE)`

**Reply by baptiste** [reply]
ideally a full reproducible example, so we wouldn't have to make one to try and understand the problem. 

WIth the example below, things behave like I would expect: can you clarify what you find problematic with the figure sizes?

`````
---
title: "test width"
format: html
---

## Quarto

Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <https://quarto.org>.

Figure with default settings
```{r figure}
#| echo: false
library(ggplot2)
theme_set(theme_minimal() +
  theme(plot.title = element_text(size = 14),
        plot.subtitle = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10),
        axis.title = element_text(size = 14),
        axis.text = element_text(size = 12),
        strip.text = element_text(size = 12),
        plot.caption = element_text(size = 10),
        panel.background = ggplot2::element_rect(color = "#F2F2F2"), 
        strip.background = ggplot2::element_rect(color = "#F2F2F2", fill = "#F2F2F2"),
        panel.spacing = unit(2, "lines"))
)
d <- data.frame(x=rep(1:10,5), y = rnorm(50), f = gl(5,10, labels = sprintf("this is column %s", 1:5)))
ggplot(d, aes(x, y)) +
  facet_grid(.~f) 
```

Now with custom settings
```{r, ref.label='figure'}
#| fig-width: 10
#| out-width: 100%
#| fig-column: screen-inset
#| echo: false
```

`````

**Reply by nikita-telkar** [reply]
Here is what I have! Setting or not setting `out.width = "100%"` isn't change the output. 
Using palmer penguins: 

```
---
title: "Untitled"
format: html
editor: source
---

```{r}

library(tidyverse)
library(palmerpenguins)
knitr::opts_chunk$set(warning = FALSE, message = FALSE, error = FALSE, cache = TRUE)


```{r font and theme, message=FALSE, error=FALSE, warning=FALSE, results='hide'}

library(showtext)
library(extrafont)

font_add_google("Jost", "Jost")

showtext_auto()

my_theme <- theme_minimal() +
  theme(plot.title = element_text(family = "Jost", size = 14),
        plot.subtitle = element_text(family = "Jost", size = 12),
        legend.title = element_text(family = "Jost", size = 12),
        legend.text = element_text(family = "Jost", size = 10),
        axis.title = element_text(family = "Jost", size = 14),
        axis.text = element_text(family = "Jost", size = 12),
        strip.text = element_text(family = "Jost", size = 12),
        plot.caption = element_text(family = "Jost", size = 10))


```{r fig.width=10, fig.height=5, out.width="100%"}

pp <- palmerpenguins::penguins

pp %>% 
  ggplot(aes(x = bill_length_mm, y = bill_depth_mm, color = species)) +
  geom_point(size = 2) +
  facet_wrap(~species) +
  my_theme

```

Left: HTML; right: inline

<img width="1440" alt="Screenshot 2023-03-15 at 2 21 34 PM" src="https://user-images.githubusercontent.com/59856969/225446744-3ed63839-534a-43c7-8d3f-70cc1467b462.png">

**Reply by baptiste** [reply]
note that it's the 'inline' (I assume you mean Rstudio inline preview of R chunks?) that has small fonts in this case, whereas in the original post above you suggested the problem was the opposite?

Are you enquiring about the font size in the Quarto output (HTML page -- in which case `out-width: 5%` definitely affects it), or in the Rstudio editor? I don't know if it's possible to adjust this at will for the latter, but it sounds like a question for the Rstudio interface more than for Quarto itself?

**Reply by nikita-telkar** [reply]
Oh, it might be! 

Sorry about that, and thank you for redirecting me, I really appreciate it! I'll check with the RStudio team.

**Reply by cscheid** [reply]
Just to add a bit more context.

I don't know what RMarkdown does in this setting, but your code cell says this:

````
```{r fig.width=10, fig.height=5, out.width="100%"}
```
````

fig.width and fig.height control the size of the image in inches (multiplied by fig-dpi), and your font size is in pixels. When you say `out.width="100%"`, you're telling quarto "ignore the size of the image, and make sure it stretches either way to fit the entire width".

These two are contradictory instructions, and quarto is "only doing what you asked it to do". If you want to control the size of the fonts in fine detail, you probably don't want to fix out-width.

**Reply by cscheid** [reply]
You can see this in action here: https://carlos-scheidegger.quarto.pub/issue-4848-repro/

Look at what the HTML inspector shows:

<img width="547" alt="image" src="https://user-images.githubusercontent.com/285675/225453344-c779fca3-0832-4be4-bd2e-1845ac2b295c.png">

In this case, your image dimensions are quite large:

<img width="253" alt="image" src="https://user-images.githubusercontent.com/285675/225453531-a54b2d26-cc65-4c80-8f98-1580dd71b730.png">

You've asked quarto to render them this way, with `fig.width="10"`. If we were to render that "in the requested size", we'd get this (you can see for yourself by disabling both the `max-width` and `width`) properties in the inspector:

<img width="1011" alt="image" src="https://user-images.githubusercontent.com/285675/225453728-6847e4e8-3846-4293-b73b-dc77f6bcde9c.png">

Note the scroll bar at the bottom. That image is more than twice as wide as the available width!

Again, I don't know how RMarkdown handles these, but in this case, quarto is "simply" doing what your .qmd asked it to.

**Reply by cderv** [reply]
In knitr, `out.width` is used to set the width to put on the HTML `width`  attribute in `img`. With Quarto and Pandoc, it seems a `style` attribute is used instead with `width` set in CSS (not as `img` attributes). When `out.width` is not set, then it is computed using `dpi` as mentioned. 

This difference is due to how Pandoc translate the width attribute to HTML
````
> pandoc --to html
![](test2_files/figure-html/plot1-1.png){width=100%}
^Z
<p><img src="test2_files/figure-html/plot1-1.png"
style="width:100.0%" /></p>
````

In R Markdown, the HTML is written directly by **knitr** in the intermediate .md file. 

Anyhow, I don't know the impact of using style over the width attribute for `img`. Probably something to do with css rules override. Not sure which is the best thing to do.  And moreover not sure it would impact the figure size in document. 

Otherwise, `fig.height` and `fig.width` should be used to configure the graphic device that save the plot file on disk. No matter `out.width`, the file shoud be identical. I'll see if we have some difference in that somehow.

**Reply by baptiste** [reply]
I _think_ the OP is asking about the inline preview of code chunks within the Rstudio editor, i.e.

![Screenshot 2023-03-16 at 12 55 54](https://user-images.githubusercontent.com/18970/225470275-cff51109-f2d1-4cd2-b2fd-ed5c866ec408.png)

(I may be wrong)

**Reply by cderv** [reply]
If so then this is something from the IDE itself, and there are already opened issue about that. See all of them in this EPIC: https://github.com/rstudio/rstudio/issues/12740

I believe there is a specific behavior in IDE for inline plot. From https://github.com/rstudio/rstudio/issues/11818
> plots within the notebook (when no figure width or height are set) try to be adaptive to the editor size

So in your example above setting `fig.width=10, fig.height=5` should change the inline figure you see. 

Anyhow, we don't really control what the IDE is doing if this is indeed the issue.

**Reply by nikita-telkar** [reply]
Thank you all - I really appreciate the time spent on this thread!

Yes, @baptiste is correct, I'm talking about the difference with the RStudio IDE ggplot figure output when using a `.qmd` and the output observed in the knitted HTML of that same `.qmd`. Have posted in the RStudio GitHub discussions.

**Reply by joekeane7** [reply]
Could you paste the link to your discussion post please @nikita-telkar? I am having the same issue and wondering if you found a solution. I only get the shrinking text when I specify the size of the element in ggplot, if you use the default sizes on a simple plot such as the below the ratio of text and plot sizes remains the same in the html output (in my case anyway). But this isn't appropriate for all plotting so wondered if you got anywhere.

library(tidyverse)
ggplot(mtcars, aes(x=wt, y=mpg)) +
  geom_point() +
  geom_smooth()
